File: scripts/capacity/fixtures/rustpbx-router.ts
import { timingSafeEqual } from 'node:crypto';
import { createServer, type IncomingMessage, type Server, type ServerResponse } from 'node:http';
import { fileURLToPath } from 'node:url';

const DEFAULT_MAX_BODY_BYTES = 1024 * 1024;
const FORWARD_ROUTES: Readonly<Record<string, string>> = Object.freeze({
  '18005550200': 'sip:uas@172.30.44.22:5060',
  '18005550201': 'sip:uas@172.30.44.23:5060',
  '18005550202': 'sip:uas@172.30.44.24:5060',
  '18005550203': 'sip:uas@172.30.44.25:5060',
  '18005550204': 'sip:uas@172.30.44.26:5060',
  '18005550205': 'sip:uas@172.30.44.27:5060;transport=tcp',
  '18005550206': 'sip:uas@172.30.44.28:5060'
});

export function createRustPbxCapacityRouter(options: {
  token: string;
  max_body_bytes?: number;
}): Server {
  const token = boundedToken(options.token);
  const maxBodyBytes = boundedMaxBodyBytes(options.max_body_bytes);
  const evidence = { router_requests: 0, cdr_requests: 0 };

  return createServer(async (request, response) => {
    try {
      const path = new URL(request.url || '/', 'http://router.invalid').pathname;
      if (request.method === 'GET' && path === '/health') {
        writeJson(response, 200, { status: 'ok' });
        return;
      }
      if (!authorized(request, token)) {
        request.resume();
        writeJson(response, 401, { error: 'unauthorized' });
        return;
      }
      if (request.method === 'GET' && path === '/evidence') {
        writeJson(response, 200, { ...evidence });
        return;
      }
      if (request.method !== 'POST') {
        request.resume();
        writeJson(response, 404, { error: 'not_found' });
        return;
      }
      if (path === '/router') {
        const body = await readBody(request, maxBodyBytes);
        if (body === null) {
          writeJson(response, 413, { error: 'body_too_large' });
          return;
        }
        const payload = parseJsonObject(body);
        const required = ['call_id', 'from', 'to', 'direction', 'method', 'uri'];
        if (!payload || required.some((field) => typeof payload[field] !== 'string' || !payload[field])) {
          writeJson(response, 422, { error: 'invalid_router_payload' });
          return;
        }
        evidence.router_requests += 1;
        const target = FORWARD_ROUTES[sipUser(String(payload.to))];
        if (target) {
          writeJson(response, 200, {
            action: 'forward', targets: [target], strategy: 'sequential',
            record: false, timeout: 30, max_ring_time: 30
          });
          return;
        }
        writeJson(response, 200, { action: 'reject', status: 486, reason: 'acceptance-route' });
        return;
      }
      if (path === '/cdr') {
        const contentType = String(request.headers['content-type'] || '');
        const body = await readBody(request, maxBodyBytes);
        if (body === null) {
          writeJson(response, 413, { error: 'body_too_large' });
          return;
        }
        if (!/^multipart\/form-data;\s*boundary=/i.test(contentType)
          || !body.includes('name="calllog.json"')) {
          writeJson(response, 422, { error: 'invalid_cdr_payload' });
          return;
        }
        evidence.cdr_requests += 1;
        writeJson(response, 200, { status: 'accepted' });
        return;
      }
      request.resume();
      writeJson(response, 404, { error: 'not_found' });
    } catch {
      if (!response.headersSent) writeJson(response, 400, { error: 'invalid_request' });
      else response.destroy();
    }
  });
}

function authorized(request: IncomingMessage, expected: string): boolean {
  const actual = String(request.headers['x-pbx-key'] || '');
  const left = Buffer.from(actual);
  const right = Buffer.from(expected);
  return left.length === right.length && timingSafeEqual(left, right);
}

function readBody(request: IncomingMessage, maximum: number): Promise<string | null> {
  return new Promise((resolve, reject) => {
    const declared = Number(request.headers['content-length'] || 0);
    let tooLarge = Number.isFinite(declared) && declared > maximum;
    let size = 0;
    const chunks: Buffer[] = [];
    request.on('data', (chunk: Buffer) => {
      size += chunk.length;
      if (size > maximum) tooLarge = true;
      if (!tooLarge) chunks.push(chunk);
    });
    request.on('end', () => resolve(tooLarge ? null : Buffer.concat(chunks).toString('utf8')));
    request.on('error', reject);
  });
}

function parseJsonObject(value: string): Record<string, unknown> | null {
  try {
    const parsed: unknown = JSON.parse(value);
    return parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed)
      ? parsed as Record<string, unknown>
      : null;
  } catch {
    return null;
  }
}

function sipUser(value: string): string {
  return /sip:([^@;>]+)/i.exec(value)?.[1] || '';
}

function writeJson(response: ServerResponse, status: number, value: object): void {
  const body = JSON.stringify(value);
  response.writeHead(status, {
    'content-type': 'application/json',
    'content-length': Buffer.byteLength(body),
    connection: 'keep-alive'
  });
  response.end(body);
}

function boundedToken(value: string): string {
  if (!value || value.length < 16 || value.length > 512) throw new Error('Router token is invalid');
  return value;
}

function boundedMaxBodyBytes(value: number | undefined): number {
  const result = value ?? DEFAULT_MAX_BODY_BYTES;
  if (!Number.isInteger(result) || result < 64 || result > 16 * 1024 * 1024) {
    throw new Error('Router body limit is invalid');
  }
  return result;
}

if (process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1]) {
  const token = String(process.env.RUSTPBX_WEBHOOK_TOKEN || '');
  const host = String(process.env.IVEKIT_CAPACITY_ROUTER_HOST || '0.0.0.0');
  const port = Number(process.env.IVEKIT_CAPACITY_ROUTER_PORT || 8081);
  if (!Number.isInteger(port) || port < 1 || port > 65535) throw new Error('Router port is invalid');
  const server = createRustPbxCapacityRouter({ token });
  server.listen(port, host, () => process.stdout.write(`iveKit capacity Router listening on ${host}:${port}\n`));
  for (const signal of ['SIGINT', 'SIGTERM'] as const) {
    process.on(signal, () => server.close(() => process.exit(0)));
  }
}

File: services/converact-service/acceptance/rustpbx-router.py
import hmac
import json
import os
import re
from email.parser import BytesParser
from email.policy import default as email_policy
from http.server import BaseHTTPRequestHandler, HTTPServer


MAX_BODY_BYTES = 1024 * 1024
WEBHOOK_TOKEN = os.environ["RUSTPBX_WEBHOOK_TOKEN"]
EVIDENCE = {"router_requests": 0, "cdr_requests": 0}
FORWARD_ROUTES = {
    "18005550200": ("sip:uas@172.30.44.22:5060", 30),
    "18005550201": ("sip:uas@172.30.44.23:5060", 30),
    "18005550202": ("sip:uas@172.30.44.24:5060", 30),
    "18005550203": ("sip:uas@172.30.44.25:5060", 30),
    "18005550204": ("sip:uas@172.30.44.26:5060", 30),
    "18005550205": ("sip:uas@172.30.44.27:5060;transport=tcp", 30),
    "18005550206": ("sip:uas@172.30.44.28:5060", 30),
}


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/health":
            self.respond(200, {"status": "ok"})
            return
        if self.path == "/evidence":
            if not self.authorized():
                self.respond(401, {"error": "unauthorized"})
                return
            self.respond(200, dict(EVIDENCE))
            return
        self.respond(404, {"error": "not_found"})

    def do_POST(self):
        if not self.authorized():
            self.respond(401, {"error": "unauthorized"})
            return
        if self.path == "/router":
            try:
                payload = self.read_json()
            except ValueError as error:
                self.respond(422, {"error": str(error)})
                return
            required = ("call_id", "from", "to", "direction", "method", "uri")
            if any(not isinstance(payload.get(name), str) or not payload[name] for name in required):
                self.respond(422, {"error": "invalid_router_payload"})
                return
            EVIDENCE["router_requests"] += 1
            destination = sip_user(payload["to"])
            route = FORWARD_ROUTES.get(destination)
            if route:
                target, max_ring_time = route
                self.respond(200, {
                    "action": "forward",
                    "targets": [target],
                    "strategy": "sequential",
                    "record": False,
                    "timeout": 30,
                    "max_ring_time": max_ring_time,
                })
                return
            self.respond(200, {"action": "reject", "status": 486, "reason": "acceptance-route"})
            return
        if self.path == "/cdr":
            try:
                self.read_call_record()
            except ValueError as error:
                self.respond(422, {"error": str(error)})
                return
            EVIDENCE["cdr_requests"] += 1
            self.respond(200, {"status": "accepted"})
            return
        self.respond(404, {"error": "not_found"})

    def read_json(self):
        try:
            payload = json.loads(self.read_body())
        except (UnicodeDecodeError, json.JSONDecodeError) as error:
            raise ValueError("invalid_json") from error
        if not isinstance(payload, dict):
            raise ValueError("invalid_json_object")
        return payload

    def read_call_record(self):
        content_type = self.headers.get("content-type", "")
        if not content_type.lower().startswith("multipart/form-data;"):
            raise ValueError("invalid_cdr_content_type")
        message = BytesParser(policy=email_policy).parsebytes(
            f"Content-Type: {content_type}\r\nMIME-Version: 1.0\r\n\r\n".encode("ascii") + self.read_body()
        )
        for part in message.iter_parts():
            if part.get_param("name", header="content-disposition") != "calllog.json":
                continue
            content = part.get_content()
            if isinstance(content, bytes):
                content = content.decode("utf-8")
            try:
                payload = json.loads(content)
            except (UnicodeDecodeError, json.JSONDecodeError) as error:
                raise ValueError("invalid_cdr_json") from error
            required_strings = ("callId", "caller", "callee")
            if not isinstance(payload, dict) or any(
                not isinstance(payload.get(name), str) or not payload[name] for name in required_strings
            ) or not isinstance(payload.get("statusCode"), int):
                raise ValueError("invalid_cdr_payload")
            return payload
        raise ValueError("missing_cdr_part")

    def read_body(self):
        raw_length = self.headers.get("content-length", "")
        if not raw_length.isdigit():
            raise ValueError("invalid_content_length")
        length = int(raw_length)
        if length < 2 or length > MAX_BODY_BYTES:
            raise ValueError("invalid_body_size")
        return self.rfile.read(length)

    def authorized(self):
        return hmac.compare_digest(self.headers.get("X-PBX-Key", ""), WEBHOOK_TOKEN)

    def log_message(self, _format, *_args):
        return

    def respond(self, status, body):
        payload = json.dumps(body, separators=(",", ":")).encode("utf-8")
        self.send_response(status)
        self.send_header("content-type", "application/json")
        self.send_header("content-length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)


def sip_user(value):
    match = re.search(r"sip:([^@;>]+)", value, re.IGNORECASE)
    user = match.group(1) if match else ""
    return user[1:] if user.startswith("+") else user


HTTPServer(("0.0.0.0", 8081), Handler).serve_forever()
